File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode node = head;
        if (head == null){
            return head;
        }
        ListNode ans = rev (head, head.next);
        head.next = null;
        return ans;
    }
    public ListNode rev(ListNode first, ListNode second){
        if (second == null)
        {
            return first;
        }
        ListNode temp = second.next;
        second.next = first;
        return rev(second, temp);
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {    
    public ListNode MergeKLists(ListNode[] lists) {
        if (lists.Length == 0) return null;
         for (int i = 1; i < lists.Length; i++) {
            lists[i] = Merge(lists[i], lists[i - 1]);
        }
        return lists[lists.Length - 1];
    }
    private ListNode Merge(ListNode l1, ListNode l2){
        ListNode dummy = new ListNode(0);
        ListNode p = dummy;

        int i = 0;
        while (l1 != null && l2 != null){
            if(l1.val < l2.val){
                p.next = l1;
                l1 = l1.next;
            }
            else{
                p.next = l2;
                l2 = l2.next;
            }
            p = p.next;
        }
        while (l1 != null){
            p.next = l1;
            p = p.next;
            l1 = l1.next;
        }
        while (l2 != null){
            p.next = l2;
            p = p.next;
            l2 = l2.next;
        }
        return dummy.next;
    }
}
